plugins/crate-copy-azure/src/main/java/io/crate/copy/azure/AzureFileInput.java

```diff
@@ -0,0 +1,147 @@
+import static io.crate.copy.azure.AzureFileOutput.resourcePath;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.apache.opendal.AsyncOperator;
+import org.apache.opendal.Entry;
+import org.apache.opendal.Operator;
+import org.elasticsearch.common.settings.Settings;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.VisibleForTesting;
+
+import io.crate.analyze.CopyStatementSettings;
+import io.crate.execution.engine.collect.files.FileInput;
+import io.crate.execution.engine.collect.files.Globs.GlobPredicate;
+
+/**
+ * File reading components operate with URI.
+ * All URI-s in the public API follow the contract "outgoing/incoming" URI is Azure compatible.
+ * This is accomplished by transforming user provided URI to the Azure compatible format only once.
+ * Outgoing URI-s are then used by other components and sent back to this component,
+ * so outgoing format (expandURI) implicitly dictates incoming URI-s format (getStream).
+ */
+public class AzureFileInput implements FileInput {
+
+    private static final Pattern HAS_GLOBS_PATTERN = Pattern.compile("^((azblob://)[^\\*]*/)[^\\*]*\\*.*");
+
+    private final Map<String, String> config;
+    private final URI uri;
+    private final GlobPredicate uriPredicate;
+    private final String preGlobPath;
+    private final Operator operator;
+    private final SharedAsyncExecutor sharedAsyncExecutor;
+
+    /**
+     * @param uri is in user provided format (azblob://path/to/dir)
+     */
+    public AzureFileInput(SharedAsyncExecutor sharedAsyncExecutor, URI uri, Settings settings) {
+        this.config = AzureBlobStorageSettings.openDALConfig(settings, CopyStatementSettings.commonCopyFromSettings);
+        this.sharedAsyncExecutor = sharedAsyncExecutor;
+        this.operator = AsyncOperator.of(NAME, config, sharedAsyncExecutor.asyncExecutor()).blocking();
+        // Pre-glob path operates with user-provided URI as GLOB pattern reflects user facing COPY FROM uri format.
+        this.preGlobPath = toPreGlobPath(uri);
+
+        String resourcePath = resourcePath(uri);
+        this.uri = URI.create(resourcePath);
+        // Glob predicate operates with normalized resource URI to reflect OpenDAL List API response format.
+        // List API returns entries without leading backslash.
+        this.uriPredicate = new GlobPredicate(resourcePath.substring(1));
+    }
+
+    /**
+     * @return List<URI> in Azure compatible format.
+     */
+    @Override
+    public List<URI> expandUri() throws IOException {
+        var operator = operator(); // Only for tests, otherwise could use this.operator
+        if (isGlobbed() == false) {
+            return List.of(uri);
+        }
+        List<URI> uris = new ArrayList<>();
+        List<Entry> entries = operator.list(preGlobPath);
```

Comment:
Is it save to pass a nullable value to `operator.list()`?

Reply:
YES, because we never get there if preGlobPath is null. 

We shortcut on `if (isGlobbed() == false)`  which is false on NULL. I will add an assertion.